filebeat/input/log/harvester.go

```diff
@@ -313,6 +313,9 @@ func (h *Harvester) Run() error {
 					},
 				},
 			}
+			if len(message.Original) > 0 {
+				fields.Put("log.original", string(message.Original))
+			}
```

Comment:
Quite unlikely but still possible that the original event has already the `log.original` present in the JSON structure, this means we would override the value.

I think we should do the following:

- Making sure that we don't override existing value.
- Maybe log if the value already exists.
- Make sure that we can configure the target field.

Reply:
Besides the given proposal, the correct behavior also depends on the JSON settings.

If `keys_under_root` is disabled we can just write to `log.original`.

If `keys_under_root` is enabled and:
- `overwrite_keys==false` => overwrite `log.original` if present. If `log` is no object, but a string remove `log` before putting `log.original`
- `overwrite_keys==true` => only write to `log.original` if not present AND `log` is an object.

Maybe it would be easier to have the json reader/parser add the field.

Reply:
The new `log.original` is a metadata field of the event. It should behave the same way as `log.offset` or `log.source` does. Both of them are added to the fields first and can be overwritten later if JSON settings require it. Thus, we have the correct behaviour you have described.
So it is not the responsibility of the JSON reader to add it. Especially if later we are adding this feature for all event formats. It is metadata like `log.source` and should be added the same way, so users can expect the same well-known behaviour.

Reply:
👍 

Do we have tests for this working with different combinations of configs? So no one breaks behavior in the future by accident.

Reply:
I only added test cases with `keys_under_root` enabled. I am creating an issue to add more tests.

Reply:
Done: https://github.com/elastic/beats/issues/9273

Reply:
Since we have the `rename` processor I would propose to remove `keys_under_root` options in 7.0 to eliminate such issues. What I haven't tested is if rename works properly when moving fields under root.